pinot-core/src/main/java/org/apache/pinot/core/plan/maker/InstancePlanMakerImplV2.java

```diff
@@ -158,7 +170,8 @@ public PlanNode makeSegmentPlanNode(IndexSegment indexSegment, QueryContext quer
     } else {
       indexSegment.prefetch(queryContext.getColumns());
     }
-    if (QueryContextUtils.isAggregationQuery(queryContext)) {
+    if (QueryContextUtils.isAggregationQuery(queryContext) ||
+            (QueryContextUtils.isDistinctQuery(queryContext) && _useDictionaryForDistinct)) {
```

Comment:
We have a branch for distinct queries (line 189 for existing code). You can check whether the query is fit for dictionary based plan similar to aggregation queries.